File: src/lib/trips.ts
import { apiFetch } from "@/lib/api";
import type { TripPlan } from "@/lib/quiz";

export interface TripStat {
  label: string;
  value: string;
}

export interface TripActivity {
  place_id: number;
  name: string;
  time: string;
  time_slot: string;
  type: string;
  desc: string;
  duration: string;
  latitude: number | null;
  longitude: number | null;
  location_confirmed: boolean;
}

export interface MapPin {
  place_id: number;
  name: string;
  day_number: number;
  time_slot: string;
  activity_type: string | null;
  latitude: number;
  longitude: number;
}

export interface DayItinerary {
  day: number;
  theme: string;
  date: string | null;
  activities: TripActivity[];
}

export interface TripDetail {
  trip_plan: TripPlan;
  destination: string;
  duration: string;
  tags: string;
  stats: TripStat[];
  itinerary: DayItinerary[];
  map_pins: MapPin[];
  geocoding_configured: boolean;
  places_geocoded: number;
  source: string;
  fallback_reason: string | null;
  is_shared?: boolean;
  share_caption?: string | null;
}

export interface TripListItem {
  id: number;
  destination: string | null;
  start_date: string | null;
  end_date: string | null;
  status: string;
  itinerary_source: string | null;
  has_itinerary: boolean;
  created_at: string;
  updated_at: string;
}

export interface TripAlternative {
  title: string;
  tagline: string;
  highlights: string[];
  budget_note: string;
  match_percent: number;
}

export interface EditTripResult {
  alternatives: TripAlternative[];
  source: string;
  fallback_reason: string | null;
}

export interface StoredChatMessage {
  id: number;
  role: "user" | "assistant";
  content: string;
  proposes_edit: boolean;
  apply_instruction: string | null;
  created_at: string;
}

export interface ChatResponse {
  message: string;
  source: string;
  fallback_reason: string | null;
  proposes_edit: boolean;
  apply_instruction: string | null;
  assistant_message_id: number | null;
  itinerary_updated: boolean;
  trip: TripDetail | null;
}

export interface DestinationSuggestion {
  city: string;
  country: string;
  blurb: string;
}

export interface SuggestDestinationsResult {
  suggestions: DestinationSuggestion[];
  source: string;
  fallback_reason: string | null;
}

export interface FlightSegment {
  origin: string;
  origin_code: string;
  destination: string;
  destination_code: string;
  departure_at: string | null;
  arrival_at: string | null;
  duration: string | null;
  airline: string;
  stops: number;
}

export interface FlightOffer {
  id: string;
  airline: string;
  price: string;
  price_note?: string | null;
  price_amount: number | null;
  currency: string;
  outbound: FlightSegment;
  inbound: FlightSegment | null;
  booking_url: string | null;
}

export interface FlightsResult {
  offers: FlightOffer[];
  search_url: string;
  source: string;
  fallback_reason: string | null;
  origin_code: string | null;
  destination_code: string | null;
}

export interface HotelOption {
  id: string;
  name: string;
  area: string;
  stars: number;
  price_per_night: string;
  price_note?: string | null;
  price_tier: string;
  amenities: string[];
  booking_url: string;
}

export interface HotelsResult {
  hotels: HotelOption[];
  search_url: string;
  source: string;
  check_in: string | null;
  check_out: string | null;
}

function normalizeTripDetail(raw: TripDetail): TripDetail {
  return {
    ...raw,
    map_pins: raw.map_pins ?? [],
    stats: raw.stats ?? [],
    geocoding_configured: raw.geocoding_configured ?? false,
    places_geocoded: raw.places_geocoded ?? 0,
    itinerary: (raw.itinerary ?? []).map((day) => ({
      ...day,
      activities: (day.activities ?? []).map((activity, index) => ({
        place_id: activity.place_id ?? index,
        name: activity.name ?? "Activity",
        time: activity.time ?? "",
        time_slot: activity.time_slot ?? "morning",
        type: activity.type ?? "Activity",
        desc: activity.desc ?? "",
        duration: activity.duration ?? "1 hr",
        latitude: activity.latitude ?? null,
        longitude: activity.longitude ?? null,
        location_confirmed: activity.location_confirmed ?? false,
      })),
    })),
  };
}

export function tripHasItineraryActivities(trip: TripDetail): boolean {
  return trip.itinerary.some((day) => day.activities.length > 0);
}

export async function suggestDestinations(tripPlanId: number): Promise<SuggestDestinationsResult> {
  return apiFetch<SuggestDestinationsResult>("/trips/suggest-destinations", {
    method: "POST",
    body: JSON.stringify({ trip_plan_id: tripPlanId }),
  });
}

export async function selectDestination(tripPlanId: number, destination: string): Promise<TripPlan> {
  return apiFetch<TripPlan>(`/trips/${tripPlanId}/destination`, {
    method: "POST",
    body: JSON.stringify({ destination }),
  });
}

export async function generateTrip(tripPlanId: number): Promise<TripDetail> {
  const result = await apiFetch<TripDetail>("/trips/generate", {
    method: "POST",
    body: JSON.stringify({ trip_plan_id: tripPlanId }),
  });
  return normalizeTripDetail(result);
}

export async function getTrip(tripPlanId: number): Promise<TripDetail> {
  const result = await apiFetch<TripDetail>(`/trips/${tripPlanId}`);
  return normalizeTripDetail(result);
}

export async function getTripFlights(tripPlanId: number): Promise<FlightsResult> {
  return apiFetch<FlightsResult>(`/trips/${tripPlanId}/flights`);
}

export async function getTripHotels(tripPlanId: number): Promise<HotelsResult> {
  return apiFetch<HotelsResult>(`/trips/${tripPlanId}/hotels`);
}

export async function enrichTripPlaces(tripPlanId: number): Promise<TripDetail> {
  const result = await apiFetch<TripDetail>(`/trips/${tripPlanId}/enrich-places`, { method: "POST" });
  return normalizeTripDetail(result);
}

const LAST_TRIP_KEY = "rihlatech_last_trip_id";
const LAST_PAGE_KEY = "rihlatech_last_page";

export function saveLastTripId(tripPlanId: number): void {
  localStorage.setItem(LAST_TRIP_KEY, String(tripPlanId));
}

export function loadLastTripId(): number | null {
  const raw = localStorage.getItem(LAST_TRIP_KEY);
  if (!raw) return null;
  const id = Number(raw);
  return Number.isFinite(id) ? id : null;
}

export function saveLastPage(page: string): void {
  localStorage.setItem(LAST_PAGE_KEY, page);
}

export function loadLastPage(): string | null {
  return localStorage.getItem(LAST_PAGE_KEY);
}

export function clearLastPage(): void {
  localStorage.removeItem(LAST_PAGE_KEY);
}

export function clearLastTripId(): void {
  localStorage.removeItem(LAST_TRIP_KEY);
  localStorage.removeItem(LAST_PAGE_KEY);
}

export async function listTrips(): Promise<TripListItem[]> {
  const result = await apiFetch<{ trips: TripListItem[] }>("/trips");
  return result.trips;
}

export async function editTrip(
  tripPlanId: number,
  prompt = "Suggest 3 alternative destinations similar to this trip.",
): Promise<EditTripResult> {
  return apiFetch<EditTripResult>(`/trips/${tripPlanId}/edit`, {
    method: "POST",
    body: JSON.stringify({ prompt }),
  });
}

export async function getChatMessages(tripPlanId: number): Promise<StoredChatMessage[]> {
  const result = await apiFetch<{ messages: StoredChatMessage[] }>(`/chat/${tripPlanId}/messages`);
  return result.messages;
}

export async function sendChatMessage(
  message: string,
  tripPlanId: number,
): Promise<ChatResponse> {
  const result = await apiFetch<ChatResponse>("/chat/message", {
    method: "POST",
    body: JSON.stringify({ trip_plan_id: tripPlanId, message }),
  });
  if (result.trip) {
    result.trip = normalizeTripDetail(result.trip);
  }
  return result;
}

export async function applyTripEdit(
  tripPlanId: number,
  instruction: string,
  chatMessageId?: number,
): Promise<TripDetail> {
  const result = await apiFetch<TripDetail>(`/trips/${tripPlanId}/apply-edit`, {
    method: "POST",
    body: JSON.stringify({
      instruction,
      chat_message_id: chatMessageId ?? null,
    }),
  });
  return normalizeTripDetail(result);
}

export async function sendConsultMessage(
  message: string,
  history: { role: "user" | "assistant"; content: string }[] = [],
): Promise<{ message: string; source: string; fallback_reason: string | null }> {
  return apiFetch("/chat/consult", {
    method: "POST",
    body: JSON.stringify({ message, history }),
  });
}

export async function deleteTrip(tripPlanId: number): Promise<void> {
  await apiFetch<void>(`/trips/${tripPlanId}`, { method: "DELETE" });
}

export async function updatePlaceLocation(
  tripPlanId: number,
  placeId: number,
  payload: { label: string; latitude: number; longitude: number; mapbox_id?: string | null },
): Promise<TripDetail> {
  const result = await apiFetch<TripDetail>(`/trips/${tripPlanId}/places/${placeId}`, {
    method: "PATCH",
    body: JSON.stringify(payload),
  });
  return normalizeTripDetail(result);
}

File: backend/app/routers/trips.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session, joinedload

from app.database import get_db
from app.dependencies.auth import get_current_user
from app.models.user import User
from app.schemas.quiz import TripPlanPublic
from app.schemas.trip import (
    ApplyEditRequest,
    EditTripRequest,
    EditTripResponse,
    FlightsResponse,
    GenerateTripRequest,
    GenerateTripResponse,
    HotelsResponse,
    SelectDestinationRequest,
    SuggestDestinationsRequest,
    SuggestDestinationsResponse,
    TripDetailResponse,
    TripListItem,
    TripListResponse,
    UpdatePlaceLocationRequest,
)
from app.services.destinations import select_destination, suggest_destinations
from app.models.trip_plan import TripPlan
from app.services.flights import search_flights
from app.services.geocoding import enrich_trip_places, mapbox_configured
from app.services.hotels import search_hotels
from app.services.apply_edit import apply_itinerary_edit
from app.services.chat import mark_chat_edit_applied
from app.services.edit import edit_trip_alternatives
from app.services.itinerary import (
    delete_trip_plan,
    generate_itinerary,
    get_trip_detail,
    list_user_trips,
    update_place_location,
)

router = APIRouter(prefix="/trips", tags=["trips"])


@router.get("", response_model=TripListResponse)
def list_trips(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> TripListResponse:
    trips = list_user_trips(db, current_user)
    items = [
        TripListItem(
            id=t.id,
            destination=t.destination,
            start_date=t.start_date,
            end_date=t.end_date,
            status=t.status,
            itinerary_source=t.itinerary_source,
            has_itinerary=bool(t.places),
            created_at=t.created_at,
            updated_at=t.updated_at,
        )
        for t in trips
    ]
    return TripListResponse(trips=items)


@router.post("/suggest-destinations", response_model=SuggestDestinationsResponse)
def suggest_trip_destinations(
    payload: SuggestDestinationsRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> SuggestDestinationsResponse:
    return suggest_destinations(db, current_user, payload.trip_plan_id)


@router.post("/generate", response_model=GenerateTripResponse)
def generate_trip_itinerary(
    payload: GenerateTripRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> GenerateTripResponse:
    return generate_itinerary(db, current_user, payload.trip_plan_id)


@router.delete("/{trip_plan_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_trip(
    trip_plan_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> None:
    delete_trip_plan(db, current_user, trip_plan_id)


@router.get("/{trip_plan_id}", response_model=TripDetailResponse)
def get_trip(
    trip_plan_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> TripDetailResponse:
    return get_trip_detail(db, current_user, trip_plan_id)


@router.get("/{trip_plan_id}/flights", response_model=FlightsResponse)
def get_trip_flights(
    trip_plan_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> FlightsResponse:
    return search_flights(db, current_user, trip_plan_id)


@router.get("/{trip_plan_id}/hotels", response_model=HotelsResponse)
def get_trip_hotels(
    trip_plan_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> HotelsResponse:
    return search_hotels(db, current_user, trip_plan_id)


@router.post("/{trip_plan_id}/enrich-places", response_model=TripDetailResponse)
def enrich_trip_places_endpoint(
    trip_plan_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> TripDetailResponse:
    trip = (
        db.query(TripPlan)
        .options(joinedload(TripPlan.places))
        .filter(TripPlan.id == trip_plan_id, TripPlan.user_id == current_user.id)
        .first()
    )
    if trip is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Trip plan not found")

    if mapbox_configured():
        for place in trip.places:
            if place.location_confirmed:
                continue
            place.latitude = None
            place.longitude = None
            place.mapbox_id = None
        db.commit()
        enrich_trip_places(db, trip, force=True)

    return get_trip_detail(db, current_user, trip_plan_id)


@router.patch("/{trip_plan_id}/places/{place_id}", response_model=TripDetailResponse)
def confirm_place_location(
    trip_plan_id: int,
    place_id: int,
    payload: UpdatePlaceLocationRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> TripDetailResponse:
    return update_place_location(
        db,
        current_user,
        trip_plan_id,
        place_id,
        label=payload.label,
        latitude=payload.latitude,
        longitude=payload.longitude,
        mapbox_id=payload.mapbox_id,
    )


@router.post("/{trip_plan_id}/apply-edit", response_model=TripDetailResponse)
def apply_trip_edit(
    trip_plan_id: int,
    payload: ApplyEditRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> TripDetailResponse:
    result = apply_itinerary_edit(db, current_user, trip_plan_id, payload.instruction)
    if payload.chat_message_id is not None:
        mark_chat_edit_applied(db, current_user, trip_plan_id, payload.chat_message_id)
    return result


@router.post("/{trip_plan_id}/edit", response_model=EditTripResponse)
def edit_trip(
    trip_plan_id: int,
    payload: EditTripRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> EditTripResponse:
    return edit_trip_alternatives(db, current_user, trip_plan_id, payload.prompt)


@router.post("/{trip_plan_id}/destination", response_model=TripPlanPublic)
def set_trip_destination(
    trip_plan_id: int,
    payload: SelectDestinationRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> TripPlanPublic:
    trip = select_destination(db, current_user, trip_plan_id, payload.destination)
    return TripPlanPublic.model_validate(trip)
